fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java

```diff
@@ -0,0 +1,273 @@
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.api.tool.attribute.v1.DynamicAttributeTool;
+import net.fabricmc.fabric.api.util.TriState;
+
+public final class ToolManagerImpl {
+	public interface Entry {
+		void setBreakByHand(boolean value);
+
+		void putBreakByTool(Tag<Item> tag, int miningLevel);
+
+		int getMiningLevel(Tag<Item> tag);
+	}
+
+	private static class EntryImpl implements Entry {
+		private Tag<Item>[] tags = new Tag[0];
+		private int[] tagLevels = new int[0];
+		private TriState defaultValue = TriState.DEFAULT;
+
+		@Override
+		public void setBreakByHand(boolean value) {
+			this.defaultValue = TriState.of(value);
+		}
+
+		@Override
+		public void putBreakByTool(Tag<Item> tag, int miningLevel) {
+			tag(tag); // Generate tag entry
+
+			for (int i = 0; i < tags.length; i++) {
+				if (tags[i] == tag) {
+					tagLevels[i] = miningLevel;
+					return;
+				}
+			}
+
+			Tag<Item>[] newTags = new Tag[tags.length + 1];
+			int[] newTagLevels = new int[tagLevels.length + 1];
+			System.arraycopy(tags, 0, newTags, 0, tags.length);
+			System.arraycopy(tagLevels, 0, newTagLevels, 0, tagLevels.length);
+			newTags[tags.length] = tag;
+			newTagLevels[tagLevels.length] = miningLevel;
+			tags = newTags;
+			tagLevels = newTagLevels;
+		}
+
+		@Override
+		public int getMiningLevel(Tag<Item> tag) {
+			for (int i = 0; i < tags.length; i++) {
+				if (tags[i] == tag) {
+					return tagLevels[i];
+				}
+			}
+
+			return -1;
+		}
+	}
+
+	private static final Map<Tag<Item>, Event<ToolHandler>> HANDLER_MAP = new HashMap<>();
+	private static final Event<ToolHandler> GENERAL_TOOLS_HANDLER = EventFactory.createArrayBacked(ToolHandler.class, ToolManagerImpl::toolHandlerInvoker);
+
+	private static final Map<Block, EntryImpl> ENTRIES = new IdentityHashMap<>();
+
+	/**
+	 * Returns a event for the tag provided, creates a new event if it does not exist.
+	 *
+	 * @param tag the tag provided for the tool
+	 * @return the event callback.
+	 */
+	public static Event<ToolHandler> tag(Tag<Item> tag) {
+		for (Map.Entry<Tag<Item>, Event<ToolHandler>> entry : HANDLER_MAP.entrySet()) {
+			if (entry.getKey().getId().equals(tag.getId())) {
+				return entry.getValue();
+			}
+		}
+
+		HANDLER_MAP.put(tag, EventFactory.createArrayBacked(ToolHandler.class, ToolManagerImpl::toolHandlerInvoker));
+		return HANDLER_MAP.get(tag);
+	}
+
+	/**
+	 * Returns a event used for every tag registered.
+	 *
+	 * @return the event callback.
+	 */
+	public static Event<ToolHandler> general() {
+		return GENERAL_TOOLS_HANDLER;
+	}
+
+	private static ToolHandler toolHandlerInvoker(ToolHandler[] toolHandlers) {
+		return new ToolHandler() {
+			@Override
+			public ActionResult isEffectiveOn(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user) {
+				for (ToolHandler toolHandler : toolHandlers) {
+					ActionResult effectiveOn = Objects.requireNonNull(toolHandler.isEffectiveOn(tag, state, stack, user));
```

Comment:
Maybe instead of hard failing here, we just return `FAIL` or `PASS`?

Reply:
I think you shouldn't be returning null anyway

Reply:
Null-checking catches errors more quickly, so I'd say it's OK.